### PR TITLE
[AMORO-1865]build a rule of relationship between table and resource group

### LINKED ISSUE #1865: [Improvement]: Build a rule of relationship between table and optimizer/resource group
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Right now, when we want to declare a table optimized by some optimizer group, we have two clear ways:
1. set default optimizer group of a catalog, and don't declare optimizer group in table properties:
image

2. declare a property of 'self-optimizing.group' in table properties (in create table or alter table statement):
image

In practice, using default optimizer group has better experience while not flexible in case that multiple groups are necessary in one catalog. Using table property provides more flexibility but sacrifice user experience and security, imagine that every table(user) needs to know the resources behind AMS and has the authority to allocate resources, this could be a disaster.

It does't seem a big deal of this because in many cases there's only one external/default optimizer group without considerations for security and isolation.  But it would be never late to have a better way to provide user experience, isolation and security for self-optimizing

### How should we improve?

**Better user experience**
users only declare relationships in one place and use them everywhere. It's a bad idea to define a property in table which means table owner must know the concepts and instances.

It's a good idea of declaring properties in optimizer group and use an extendable rule like regex

**Better security**
Relationships of table and resource should be certain and can not be modified without the permission of the owner of resources. It is clear that  declaring properties in optimizer group fulfills this criterion

**Better isolation**
when we declare relationships of table and resources or modify them, the rules must be mutually exclusive

In conclusion, I proposed that declaring regex rules in optimizer group defines relationships of table and resources. For example:

> catalog1.db1.*
> catalog2.*.*

leads to a clear definition that this optimizer groups could be used in these tables and only used by them.

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Aireed**
## Why are the changes needed?

Close #1865.

## Brief change log

- Front-end: 
  - Configure by adding rules in the properties of optimize-group with key 'match.rules'
  
image

- Backend:
  - 1. When creating a group, rule conflict detection will be performed.
And traverse the table corresponding to the catalog default value of the current optimize-group to see if it matches the current rule and assign the value.
  - 2. When updating the optimize-group rules, rule conflict detection will be performed.
	- 1. First, the table under the current rule will be judged. If the rule no longer matches after the rule is changed, the default group of the catalog to which the table belongs will be used.
	- 2. Traverse the table corresponding to the catalog default value of the current optimize-group to see if it matches the current rule and assign the value.
  - 3. when deleting the group
	- 1. Configure the tables under the group to the default group of the catalog to which they belong.
	When the group of the above table is changed, the tableConfigureChange responsibility chain call of the table will be triggered.
  - 4. When synchronizing the external catalog, the group will be automatically matched according to regular rules.
  - 5. TableRuntimeRefreshExecutor no longer pays attention to optimize-group when refreshing.

- Special note
 We support manually adding the table name to the rules of a group, which is equivalent to the original manual configuration of the group. Because in our practice, there are some situations that require manual specification.

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: amoro-common/src/main/java/org/apache/amoro/resource/ResourceGroup.java [hunk 1]
```diff
@@ -52,6 +77,81 @@ public String getContainer() {
...
+
+  private boolean match(Map<String, String> properties, TableIdentifier tableIdentifier) {
+    String rulePropertyKey = OPTIMIZE_GROUP_RULE_MATCH_KEY;
+    if (properties != null && properties.containsKey(rulePropertyKey)) {
+      List<String> matchRules =
+          Arrays.stream(properties.getOrDefault(rulePropertyKey, "").split(","))
+              .collect(Collectors.toList());
+
+      final String tableName =
+          String.format(
```
Comment (XBaith) [inline amoro-common/src/main/java/org/apache/amoro/resource/ResourceGroup.java:133]: tableIdentifier.toString()

File: amoro-common/src/main/java/org/apache/amoro/ServerTableIdentifier.java [hunk 2]
```diff
@@ -123,6 +123,10 @@ public String toString() {
     return String.format("%s.%s.%s(tableId=%d)", catalog, database, tableName, id);
   }
 
+  public String getTableFullName() {
```
[the remark this answers is not in the document]
Comment (Aireed) [reply amoro-common/src/main/java/org/apache/amoro/ServerTableIdentifier.java:126]: > Is this method nessessary? How about invoke `getIdentifier().toString()` to get full name?

`getIdentifier().toString()` returns `TableIdentifier[catalog.db.table]`, So it's not suitable.

[the remark this answers is not in the document]
Comment (XBaith) [reply amoro-common/src/main/java/org/apache/amoro/ServerTableIdentifier.java:126]: Okay, It not `org.apache.amoro.table.TableIdentifier`. I perfer use the following method, WDYT?
```
  public org.apache.amoro.table.TableIdentifier getTableIdentifier() {
    return org.apache.amoro.table.TableIdentifier.of(catalog, database, tableName);
  }
```

**Comment by baiyangtx** [comment]
This implementation method will be contrary to the current advanced issue of splitting the rest service and the scheduling service. It makes the update of the group have to occur at the master node.

**Comment by Aireed** [comment]
we may reopen it after #2810 finished

**Pull request closed without merging**